From 3876ecfdfe9f1e87aaee6b1949e2e4da9d9b5d67 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 13 Jan 2025 15:13:17 +0800
Subject: [PATCH 087/107] mtk: mac80211: pass valid links to mt7996 when update
 vif's links in mac80211

MAC80211 used to pass active links to mt7996 when updating vif's links,
and mt7996 add/delete link according to the links provided by MAC80211.

However, we expect links being added/deleted only when valid links
changes, not active links.

This commits ass valid links to mt7996 when update vif's links in mac80211

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 net/mac80211/link.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/net/mac80211/link.c b/net/mac80211/link.c
index 85eae892..e05ed78b 100644
--- a/net/mac80211/link.c
+++ b/net/mac80211/link.c
@@ -199,7 +199,6 @@ static int ieee80211_vif_update_links(struct ieee80211_sub_if_data *sdata,
 				      u16 new_links, u16 dormant_links)
 {
 	u16 old_links = sdata->vif.valid_links;
-	u16 old_active = sdata->vif.active_links;
 	unsigned long add = new_links & ~old_links;
 	unsigned long rem = old_links & ~new_links;
 	unsigned int link_id;
@@ -278,10 +277,13 @@ static int ieee80211_vif_update_links(struct ieee80211_sub_if_data *sdata,
 		ieee80211_set_vif_links_bitmaps(sdata, new_links, dormant_links);
 
 		/* tell the driver */
-		ret = drv_change_vif_links(sdata->local, sdata,
-					   old_links & old_active,
-					   new_links & sdata->vif.active_links,
-					   old);
+		/* FIXME
+		 * MLMR solution does link add/del when valid links change
+		 * and so we change to report old/new valid links to driver.
+		 * Also, TTLM is handled in other callbacks (not upstream yet).
+		 */
+		ret = drv_change_vif_links(sdata->local, sdata, old_links,
+					   new_links, old);
 		if (!new_links)
 			ieee80211_debugfs_recreate_netdev(sdata, false);
 	}
-- 
2.45.2

